mitodl/ocw-studio
# add local s3 storage emulation with minio

#### Pre-Flight checklist


#### What are the relevant tickets?
Closes https://github.com/mitodl/ocw-studio/issues/1420

#### What's this PR do?
This rather large PR adds new docker containers to the local docker-compose configuration to run [Minio](https://github.com/minio/minio) locally. Minio emulates Amazon S3, allowing you to isolate the `ocw-studio` publishing process locally for testing without having to touch any S3 buckets actually in AWS. Any initialization of the `boto3` library in Python code will now point itself at Minio if `settings.ENVIRONMENT` is set to "dev." Management commands that upsert pipelines to Concourse also now use a new utility function to strip out lines based on if your environment is "dev" or not. Dev pipelines will override the AWS endpoint URL to point to Minio and skip things like the Slack webhooks, Open Discussions webhooks, automatic triggering, etc.

#### How should this be manually tested?
 - Read the updated readme and set up the Minio env variables based on it. Also look at the updated Concourse section and the readme overall and make sure the changes make sense. Also, make sure you have Github integration set up with a custom organization.
 - Spin up `ocw-studio` with `docker-compose up`
 - Pull and restore a backup of the production database:
```
pg_dump PROD_DB_URL > ./dumps/prod_TODAY_DATE.sql
docker-compose rm -sv db
docker-compose up -d db
docker-compose exec -T -e PGPASSWORD=postgres db psql -h db -U postgres < ./dumps/prod_TODAY_DATE.sql
```
(alternatively, make sure you have an `ocw-www` site as well as an `ocw-course` site)
 - In the logs, you will see the `create-buckets` container spin up, create some containers and set permissions and then exit. This is normal and runs to create buckets in Minio.
 - Upsert the theme assets pipeline to your local Concourse with `./manage.py upsert_theme_assets_pipeline, log into Concourse at http://concourse:8080, start a theme assets build and wait for it to succeed.
 - Browse to your `ocw-www` site at http://localhost:8043/sites/ocw-www. Publish the site to draft.
 - Browse to http://localhost:8080 and log into Concourse. Find the `ocw-www` live pipeline and make sure it completes successfully.
 - Once the pipeline is complete, visit http://localhost:8044 and verify you see the OCW home page. Repeat the last 3 steps for live and verify the same thing but at http://localhost:8045
 - Create a site using the `ocw-course` starter. Visit the Resources section, click the Google Drive link and upload an example file. Click the "Sync w/ Google Drive" button and wait until it succeeds. Visit http://localhost:9001 and login using the credentials detailed in the readme. Browse to the storage bucket and verify that your file was successfully uploaded.
 - Go back to your `ocw-course` site and publish it to draft.
 - Go back to Concourse and locate the draft pipeline for your course site, make sure it completes.
 - Once the pipeline is complete, visit http://localhost:8044/courses/your_course and verify you see the OCW home page. Repeat the last 3 steps for live and verify the same thing but at http://localhost:8045/courses/your_course
 - Smoke test any other functionality in `ocw-studio` that you can

#### Any background context you want to provide?
This is obviously a pretty big PR that touches a lot of existing functionality, so please reach out with any questions while reviewing.
